File: mod-common/src/main/java/com/dreamdisplays/net/SyncPacket.java
package com.dreamdisplays.net;

import com.dreamdisplays.PlatformlessInitializer;

import java.util.UUID;
import net.minecraft.core.UUIDUtil;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.Identifier;
import org.jspecify.annotations.NullMarked;

// Packet for synchronizing the playback state of a display
@NullMarked
public record SyncPacket(UUID id, boolean isSync, boolean currentState, long currentTime, long limitTime) implements CustomPacketPayload {
    public static final CustomPacketPayload.Type<SyncPacket> PACKET_ID =
            new CustomPacketPayload.Type<>(Identifier.fromNamespaceAndPath(PlatformlessInitializer.MOD_ID, "sync"));

    public static final StreamCodec<FriendlyByteBuf, SyncPacket> PACKET_CODEC =
            StreamCodec.of(
                    (buf, packet) -> {
                        UUIDUtil.STREAM_CODEC.encode(buf, packet.id());
                        ByteBufCodecs.BOOL.encode(buf, packet.isSync());
                        ByteBufCodecs.BOOL.encode(buf, packet.currentState());
                        ByteBufCodecs.VAR_LONG.encode(buf, packet.currentTime());
                        ByteBufCodecs.VAR_LONG.encode(buf, packet.limitTime());
                    },
                    (buf) -> {
                        UUID id = UUIDUtil.STREAM_CODEC.decode(buf);

                        boolean isSync = ByteBufCodecs.BOOL.decode(buf);
                        boolean currentState = ByteBufCodecs.BOOL.decode(buf);
                        long currentTime = ByteBufCodecs.VAR_LONG.decode(buf);
                        long limitTime = ByteBufCodecs.VAR_LONG.decode(buf);

                        return new SyncPacket(id, isSync, currentState, currentTime, limitTime);
                    });

    @Override
    public Type<? extends CustomPacketPayload> type() {
        return PACKET_ID;
    }
}

File: common/src/main/kotlin/com/dreamdisplays/net/common/SyncPacket.kt
package com.dreamdisplays.net.common

import com.dreamdisplays.net.common.helpers.createType
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.codec.StreamCodec.of
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.network.protocol.common.custom.CustomPacketPayload.Type
import java.util.*

data class SyncPacket(
    val uuid: UUID,
    val isSync: Boolean,
    val currentState: Boolean,
    val currentTime: Long,
    val limitTime: Long,
) : CustomPacketPayload {
    override fun type(): Type<out CustomPacketPayload> = PACKET_ID

    companion object {
        val PACKET_ID: Type<SyncPacket> = createType("sync")
        val PACKET_CODEC: StreamCodec<FriendlyByteBuf, SyncPacket> = of(
            { buf, packet ->
                buf.writeUUID(packet.uuid)
                buf.writeBoolean(packet.isSync)
                buf.writeBoolean(packet.currentState)
                buf.writeVarLong(packet.currentTime)
                buf.writeVarLong(packet.limitTime)
            },
            { buf ->
                SyncPacket(
                    buf.readUUID(),
                    buf.readBoolean(),
                    buf.readBoolean(),
                    buf.readVarLong(),
                    buf.readVarLong()
                )
            }
        )
    }
}
